Zero-fee commitments using v3 transactions (feature 40/41) (#1228)
lightningnetwork/lightning-rfc: 05-onchain.md

funding_pubkey.
4. _remote node's anchor output_: one output paying to the *remote node's*
funding_pubkey.
5. _shared ephemeral anchor_: one output paying to the *local* or *remote*

> node's funding_pubkey.

Reviewer: * The ephemeral in the output name isn't used anywhere else, probably just "shared anchor"?
* Description doesn't match the fact that the shared anchor is using P2A

Reviewer 2: This wasn't fully addressed.  It still refers to a keyed anchor instead of P2A.

        - otherwise:
          - MUST broadcast the *last commitment transaction*, for which it has a
          signature, to perform a *unilateral close*.

>           - MUST spend any `to_local_anchor` or `shared_anchor` output, providing sufficient fees
>           as incentive to include the commitment transaction in a block. When using `nVersion=2`,
>           special care must be taken when spending to a third-party, because this re-introduces
>           the vulnerability that was addressed by adding the CSV delay to the non-anchor outputs.

Reviewer 2: Bringing up `nVersion=2` here is a bit confusing and may give the reader the idea they can set `nVersion=3` for `to_local_anchor` spends to avoid this issue (which won't relay).

```suggestion
          - MUST spend any `to_local_anchor` or `shared_anchor` output, providing sufficient fees
          as incentive to include the commitment transaction in a block. For `to_local_anchor`,
          special care must be taken when spending to a third-party, because this re-introduces
          the pinning vulnerability that was addressed by adding the CSV delay to the non-anchor outputs.
```

  using the revocation private key.
  - SHOULD extract the payment preimage from the transaction input witness, if
  it's not already known.

>   - if `SIGHASH_SINGLE|SIGHASH_ANYONECANPAY` is used for HTLC transactions:

Reviewer 2: ```suggestion
  - if `SIGHASH_SINGLE|SIGHASH_ANYONECANPAY` is used for HTLC transactions (i.e. `option_anchors` or `zero_fee_commitments` applies):
```

standard size limit because of the 483 HTLC-per-party limit (see
[BOLT #2](02-peer-protocol.md#the-open_channel-message)).

> Note: when `SIGHASH_SINGLE|SIGHASH_ANYONECANPAY` is used for HTLC transactions,
> the cheating node can pin spends of its HTLC-timeout/HTLC-success outputs.

Reviewer 2: ```suggestion
Note: when `SIGHASH_SINGLE|SIGHASH_ANYONECANPAY` is used for HTLC transactions (i.e. `option_anchors` or `zero_fee_commitments`),
the cheating node can pin spends of its HTLC-timeout/HTLC-success outputs.
```

  using the revocation private key.
  - SHOULD extract the payment preimage from the transaction input witness, if
  it's not already known.
  - if `SIGHASH_SINGLE|SIGHASH_ANYONECANPAY` is used for HTLC transactions
  (i.e. `option_anchors` or `zero_fee_commitments` applies):
    - MAY use a single transaction to *resolve* all the outputs.
    - If `zero_fee_commitments` applies:
      - MUST NOT exceed the 10kvB limit of v3 transactions, otherwise the
        transaction may not reach miners (as explained in the section

>         about [v3 limitations](#limitations-of-v3-transactions) below).

Author: I'm not sure I understand your comment here. HTLC transactions are using v3, so if you batch a bunch of HTLC transactions together (using the fact that their input uses `SIGHASH_SINGLE | SIGHASH_ANYONECANPAY`), the resulting transaction is a batched HTLC transaction that is using v3. So regardless of whether the commitment transaction is confirmed or not, the batched HTLC transaction cannot exceed 10kvB. Does that make sense?

# Generation of HTLC Transactions

> If `SIGHASH_ALL` is used for HTLC transactions, then HTLC-timeout and
> HTLC-success transactions are complete transactions with (hopefully!)
> reasonable fees and must be used directly.

Reviewer 2: This section describes *when* to use `SIGHASH_ALL` vs `SIGHASH_SINGLE|SIGHASH_ANYONECANPAY`.  So we shouldn't assume the reader already knows which sighashes to use for each channel type.

```suggestion
If `option_anchors` and `zero_fee_commitments` do not apply, then HTLC-timeout and
HTLC-success transactions are signed with SIGHASH_ALL and are complete transactions with (hopefully!)
reasonable fees and must be used directly.
```
